File: flutter_app/lib/models/swap_request.dart
class SwapRequest {
  SwapRequest({
    required this.id,
    required this.status,
    this.message,
    this.otherUserName,
    this.createdAt,
  });

  final int id;
  final String status;
  final String? message;
  final String? otherUserName;
  final String? createdAt;

  factory SwapRequest.fromJson(Map<String, dynamic> json, {required bool inbox}) {
    return SwapRequest(
      id: _asInt(json['id']),
      status: (json['status'] ?? 'pending') as String,
      message: json['message'] as String?,
      otherUserName: (inbox ? json['from_user_name'] : json['to_user_name']) as String?,
      createdAt: json['created_at'] as String?,
    );
  }
}

int _asInt(dynamic value) {
  if (value is int) {
    return value;
  }
  return int.tryParse('$value') ?? 0;
}

File: frontend/src/models/swapRequest.ts
import {
  asInt,
  asNullableInt,
  asNullableString,
  asRecord,
  asString,
} from './parse';

export interface SwapRequest {
  id: number;
  fromUserId: number;
  toUserId: number;
  status: string;
  message?: string;
  otherUserName?: string;
  teachSkillName?: string;
  learnSkillName?: string;
  proposedTime?: string;
  conversationId?: number;
  createdAt?: string;
  updatedAt?: string;
}

export function parseSwapRequest(value: unknown, inbox: boolean): SwapRequest {
  const json = asRecord(value);
  return {
    id: asInt(json.id),
    fromUserId: asInt(json.from_user_id),
    toUserId: asInt(json.to_user_id),
    status: asString(json.status, 'pending'),
    message: asNullableString(json.message),
    otherUserName: asNullableString(inbox ? json.from_user_name : json.to_user_name),
    teachSkillName: asNullableString(json.teach_skill_name),
    learnSkillName: asNullableString(json.learn_skill_name),
    proposedTime: asNullableString(json.proposed_time),
    conversationId: asNullableInt(json.conversation_id),
    createdAt: asNullableString(json.created_at),
    updatedAt: asNullableString(json.updated_at),
  };
}
